spock-core/src/main/java/org/spockframework/runtime/Scheduler.java

```diff
@@ -0,0 +1,67 @@
+package org.spockframework.runtime;
+
+import org.junit.runners.model.RunnerScheduler;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class Scheduler {
+  private final RunnerScheduler runnerScheduler;
+  private final boolean forceSequential;
+  private final AtomicInteger scheduledTasksCount = new AtomicInteger(0);
+  private final Semaphore concurrentJobsSemaphore = new Semaphore(0);
+  private State state = State.SCHEDULING;
+
+  public Scheduler(RunnerScheduler runnerScheduler, boolean forceSequential) {
+    this.runnerScheduler = runnerScheduler;
+    this.forceSequential = forceSequential;
+  }
+
+  public void schedule(final Runnable task) {
+    synchronized (this){
```

Comment:
Can we use a more efficient construct than `synchronized`, e.g. ReadWriteLock. Either way you should not use `this` for synchronized, but a private `Object lock = new Object()` to prevent external parties to use your lock.